# Smell: “Missing the Mask of Invalid Value”


## Motivation  
Applying a logarithm to tensors that may contain zeros or negative values produces -inf or nan, which can  

- explode the loss,  
- stall or diverge training,  
- make gradients undefined.

The standard TensorFlow safeguard is to **mask or clip** the argument before taking the logarithm:

```python
# clipping
tf.math.log(tf.clip_by_value(x, eps, 1.0))

# masking
tf.math.log(tf.where(x > 0, x, eps))
```

## Detection strategy (high-level view)

1. **AST parsing**  
   The source file is transformed into its Abstract Syntax Tree so each
   constructor call can be inspected structurally.

2. **Locate logarithm calls**  
   • A node is considered a log call when: 
   • it is an ast.Call, **and**
   • its fully-qualified name is `tf.log` or `tf.math.log`.

3. **Check the first argument**  
   The call is deemed **safe** when `call.args[0]` is itself:

   - a `tf.clip_by_value` call, or
   - a `tf.where / tf.math.where` call whose predicate enforces positivity.

4. **Decision rule**  
   The detector emits  
    REPORT: REPORT: Missing mask for tf.log at line `n`  
    where `n` is is the line number of the unsafe log call.

---

## Practical meaning for the developer  
Always guard the logarithm with a mask or clip:

```python
# BEFORE  – will trigger the smell
loss = tf.reduce_mean(tf.math.log(x))

# AFTER   – numerically safe
eps  = 1e-8
loss = tf.reduce_mean(tf.math.log(tf.clip_by_value(x, eps, 1.0)))

```

Advantages of being explicit:

- prevents `nan/inf` explosions,
- ensures gradients stay finite,
- documents the intended numeric range for future maintainers.

## Limitations

- Only `tf.log` and `tf.math.log` are checked; other aliases (`tf.math.log1p`, custom wrappers) are ignored.  
- The rule recognises masking via `tf.clip_by_value` or `tf.where only`; bespoke helpers (e.g. `safe_log(x)`) evade detection.  
- Static analysis cannot know the runtime distribution of a tensor; it may raise false positives when the developer can prove strict positivity by construction.